### PR TITLE
lazy load query filters and extract them into classes

### PR DESCRIPTION
**Pull request by ulferts**
This allows us to only load the values we actually need instead of
loading all the values regardless of whether we need them. This is
especially usefull in the api where we currently never need those
values.

Note, that this is a glitch in my opinion as we do not check whether the
filter value is valid.

PR: https://community.openproject.com/work_packages/23891/activity

### CODE REVIEW

**Comment by ulferts** [comment]
The `ToDo` codeclimate is complaining about existed before. It was only moved.

**Review by oliverguenther — CHANGES_REQUESTED** [review]
When visiting the work package list and adding an assignee, I get the following error.

```
NameError (wrong constant name group):

app/models/queries/work_packages/filter/group_filter.rb:38:in `available?'
app/models/queries/work_packages/available_filter_options.rb:59:in `block in initialize_work_package_filter'
app/models/queries/work_packages/available_filter_options.rb:59:in `reject'
app/models/queries/work_packages/available_filter_options.rb:59:in `initialize_work_package_filter'
app/models/queries/work_packages/available_filter_options.rb:35:in `block in available_work_package_filters'
app/models/queries/work_packages/available_filter_options.rb:34:in `each'
app/models/queries/work_packages/available_filter_options.rb:34:in `available_work_package_filters'
app/helpers/queries_helper.rb:166:in `fix_field_array'
app/helpers/queries_helper.rb:123:in `fields_from_params'
app/helpers/queries_helper.rb:42:in `add_filter_from_params'
app/controllers/api/experimental/concerns/query_loading.rb:72:in `prepare_query'
app/controllers/api/experimental/concerns/query_loading.rb:47:in `init_query'
app/controllers/api/experimental/work_packages_controller.rb:65:in `load_query'
app/middleware/reset_current_user.rb:47:in `call'
```

File: app/controllers/api/experimental/queries_controller.rb [hunk 1]
```diff
@@ -155,11 +150,24 @@ def authorize_update_on_query
         actions.delete(:update) if changed.empty?
       end
 
-      allowed = actions.map(&:to_sym)
+      allowed = actions
+                .map(&:to_sym)
```
Comment (oliverguenther) [inline app/controllers/api/experimental/queries_controller.rb]: Why mapping twice when we can just `...allowed?(original_query, action.to_sym)`?

**Comment by oliverguenther** [comment]
👍 👍  for the contained refactoring. It introduces legibility into that whole area.

**Comment by ulferts** [comment]
@oliverguenther thanks for spotting those constant problems. Still don't know why they where caused but letting ruby search for the constants from root solved the problem.

**Review by oliverguenther — APPROVED** [review]

**Pull request merged**